pkg/solana/txm/txm_integration_test.go

```diff
@@ -185,3 +189,185 @@ func createTransaction(ctx context.Context, t *testing.T, client *solanaClient.C
+		require.NoError(t, err)
+
+		// Create TX and wait for it to be confirmed
+		const amount = 1 * solana.LAMPORTS_PER_SOL
+		txID := "reorg-test-tx"
+		tx, lastValidBlockHeight := createTransaction(ctx, t, cl, senderPubKey, receiverPubKey, amount, true)
+		require.NoError(t, txmInstance.Enqueue(ctx, "", tx, &txID, lastValidBlockHeight))
+		require.Eventually(t, func() bool {
+			status, errGetStatus := txmInstance.GetTransactionStatus(ctx, txID)
+			if errGetStatus != nil {
+				return false
+			}
+			if status == types.Unconfirmed {
+				pTx, errPtx := txmInstance.getPendingTx(txID)
+				if errPtx != nil || len(pTx.signatures) == 0 {
+					return false
+				}
+
+				sigStatus, errStat := cl.SignatureStatuses(ctx, pTx.signatures)
+				if errStat != nil || len(sigStatus) == 0 || sigStatus[0] == nil {
+					return false
+				}
+				return sigStatus[0].ConfirmationStatus == rpc.ConfirmationStatusConfirmed
+			}
+			return false
+		}, 60*time.Second, 1*time.Second, "Transaction should reach Confirmed status")
+
+		// Simulate reorg: kill current validator and restart validator with backuped ledger before the tx.
+		// we want ledger as provided, omit --reset
+		require.NoError(t, cmd.Process.Kill())
+		_ = cmd.Wait()
+		require.NoError(t, os.RemoveAll(ledgerDir))
+		require.NoError(t, copyDir(cleanLedgerBackupDir, ledgerDir))
+		startValidator(t, ledgerDir, port, faucetPort, false)
+
+		// Check tx is not finalized yet and reorg is detected
+		status, errGetStatus := txmInstance.GetTransactionStatus(ctx, txID)
+		require.NoError(t, errGetStatus)
+		require.NotEqual(t, types.Finalized, status, "tx should not be finalized after reorg")
+		reorgLogs := obs.FilterMessageSnippet("re-org detected for transaction").Len()
+		require.Equal(t, reorgLogs, 1, "Re-org should be detected")
+		rebroadcastReorgLogs := obs.FilterMessageSnippet("confirmed re-orged tx was rebroadcasted successfully").Len()
+		require.Equal(t, rebroadcastReorgLogs, 1, "re-org tx should be rebroadcasted with new blockhash")
+
+		// Wait rebroadcasted tx to be finalized and check final balances
+		require.Eventually(t, func() bool {
+			finalStatus, errAgain := txmInstance.GetTransactionStatus(ctx, txID)
+			if errAgain != nil {
+				return false
+			}
+			return finalStatus == types.Finalized
+		}, 120*time.Second, 5*time.Second, "tx should finalize again after reorg handling")
+		finalSenderBalance, err := cl.Balance(ctx, senderPubKey)
+		require.NoError(t, err)
+		finalReceiverBalance, err := cl.Balance(ctx, receiverPubKey)
+		require.NoError(t, err)
+		require.Less(t, finalSenderBalance, initSenderBalance, "Sender balance should decrease after re-finalization")
+		require.Equal(t, amount, finalReceiverBalance, "Receiver should receive transferred amount after re-finalization")
+		status, errGetStatus = txmInstance.GetTransactionStatus(ctx, txID)
+		require.NoError(t, errGetStatus)
+		require.Equal(t, types.Finalized, status, "tx should be finalized after reorg")
+	})
```

Comment:
Could you also include a test case where we detect a re-org from `Processed` state?

Reply:
I couldn't achieve this because the local live validator transitions to `Confirmed` instantly. I can try spamming a lot of txes and see if I can overwhelm it so it delays the transition and gives a window to stop it

Reply:
Ah gotcha. In that case I think this is ok. Your mock testing for that scenario might be enough